### DISCUSSION TITLE
Understanding actions in external converters

### DISCUSSION THREAD

**Discussion by grafalex82** [Koenkk/zigbee2mqtt#9282] [Devices]
Hi All,

I am working on yet another 2-buttons switch, and currently developing an external converter. Among other features I am handling genMultistateInput cluster and would like to send single/double/tripple/whatever actions for my 2 buttons. The device shall generate separate actions for each of the buttons (e.g. single_button_1 or double_button_2), and at the same time, later I am going to add handling for pressing both buttons simultaneously (e.g. single_both).

Although it is basically working, I would like to get better understanding on what the action is. 

Questions:
1) Should actions be defined on the topmost level ([device structure](https://github.com/grafalex82/hellozigbee/blob/hello_zigbee_custom_cluster/myswitch.js#L234))

```
const device = {
    zigbeeModel: ['Hello Zigbee Switch'],
...
    exposes: [ e.action(['*_single', '*_double', '*_triple', '*_hold', '*_release'])].concat(genEndpoints(2)),
...
};

```
or it can/should be defined using .withEndpoint() altering (e.g. [while defining a specific endpoint capabilities](https://github.com/grafalex82/hellozigbee/blob/hello_zigbee_custom_cluster/myswitch.js#L158))?

```
function genEndpoint(epName) {
    return [
        e.switch().withEndpoint(epName),
...
        e.action(['single', 'double', 'triple', 'hold', 'release']).withEndpoint(epName)
    ]
}
```

2) [The official z2m external converter example](https://github.com/Koenkk/zigbee2mqtt.io/blob/master/docs/externalConvertersExample/freepad_ext.js#L83) uses wildcard syntax.

```
    exposes: [e.battery(), e.action([
        'button_*_hold', 'button_*_single', 'button_*_double', 'button_*_triple', 'button_*_quadruple', 'button_*_release'])],

```
Is there a documentation explaining this? I have not found a code that processes a wildcard symbols.

3) My device will have also non-switch endpoints. Will this wildcards generate actions for non-switch endpoints?

4) The most interesting question: what is this action definition for? Perhaps I am doing something wrong... I tried adding e.action(), and removing it completely - my HA start seeing actions only after they triggered once, but not just by definition. And adding/removing e.action() does not block z2m trigger actions. What is the purpose of e.action()?

Here is [the full code](https://github.com/grafalex82/hellozigbee/blob/hello_zigbee_custom_cluster/myswitch.js), just in case

**Comment by Koenkk** [comment]
1. topmost level
2/3. The whole expose stuff was added later, when adding this I did not know all the possible action of a device so I put a `*` there. There is no specific logic for this `*`. For newly added devices, the wildcard is not accepted anymore.
4. It is mainly used for the docs, external application like the z2m homebridge plugin also use it

**Reply by grafalex82** [reply]
Hi Koen, thank you very much for your reply. 

As for 2 and 3. Should I expose exact list of actions for all my endpoints and actions? e.g. single_button_1, single_button_2, double_button_1, double_button_2, etc

Can I expose several e.actions() so that they be concatenated by z2m?

**Reply by Koenkk** [reply]
> Should I expose exact list of actions for all my endpoints and actions? 

Yes, that is correct

> Can I expose several e.actions() so that they be concatenated by z2m?

No, just add one e.action with all the actions in it.

**Reply by grafalex82** [reply]
I see, thank you very much. 

Still confused with why Home Assistant ignores actions list, but looks like this is not a z2m issue.
